Add more descriptions & examples to the readme (#316)
canonical/microcluster: README.md

# Microcluster
## Introduction

Microcluster is a Go library that provides an easy to use framework for creating and managing highly available [dqlite](https://github.com/canonical/dqlite) clusters. It provides an extensible API and database which can be used to directly connect to your underlying services.

> Extend Microcluster with your own API endpoints, schema extensions, lifecycle actions, and additional network listeners.

Reviewer: schema and API extensions?

## Import Microcluster

Get the latest LTS release of Microcluster:

```
$ go get github.com/canonical/microcluster/v2@latest
```

## Usage
For a step-by-step practical example of creating a Microcluster project, check out the [example package](example). Examine the code or build and test it out with the included Makefile.

All of Microcluster's state is stored in the `StateDir` directory. Define a new Microcluster app:

```go
m, err := microcluster.App(microcluster.Args{StateDir: "/path/to/state"})
if err != nil {
    // ...
}
```

Extensions to Microcluster must be defined in `DaemonArgs`:
```go

dargs := microcluster.DaemonArgs{

>     Version: "0.0.0",

Reviewer: Maybe have `0.0.1` which is a "valid" first version?

    ExtensionsSchema: []schema.Update{
        // Sequential list of modifications to the database.
    },

    APIExtensions: []string{
        // Sequential list of API capabilities.
    },

    Servers: map[string]rest.Server{
        "my-server": {
            CoreAPI: true,
            ServeUnix: true,
            Resources: []rest.Resources{
                {
                    PathPrefix: "1.0",
                    Endpoints: []rest.Endpoint{
                        // Additional API endpoints to offer on "/1.0".
                    },
                },
            }
        }
    }
}
```

> Start the daemon with your configuration:

Reviewer: Worth adding a comment that this will not block but instead calls the API to tell the daemon to start? 

Maybe put a `<-exit` or something equal.

Author: `app.Start` does indeed block, there's no API to tell the daemon to start because there's nothing to serve the API before the daemon starts.

[Here's it used in the example package](https://github.com/canonical/microcluster/blob/4d80df396e335bf26f9895956e846e082bb8f624/example/cmd/microd/main.go#L204)

Reviewer: You are right, nvm then.

```go
err := m.Start(ctx, dargs)
if err != nil {
    // ...
}
```

### Define lifecycle actions (Hooks):
```go
dargs.Hooks = &state.Hooks{
    OnStart: func(ctx context.Context, s state.State) error {
        // Execute code on each startup.
    }
}
```

### Query the dqlite database directly:
```go

> _, batch, err := m.Sql(ctx, "select name, address from core_cluster_members where role='voter'")

Reviewer: Worth using the uppercase representation for the SQL syntax e.g. `SELECT` and `WHERE`.

if err != nil {
    // ...
}

for i, result := range batch.Results {
    fmt.Printf("Query %d:\n", i)

    fmt.Println("Type:", result.Type)
    fmt.Println("Columns:", result.Columns)
    fmt.Println("RowsAffected:", result.RowsAffected)
    fmt.Println("Rows:", result.Rows)
}

```

### Create your own API endpoints:
```go
endpoint := rest.Endpoint{
    Path: "mypath" // API is served over /mypath

    Post: rest.EndpointAction{Handler: myHandler} // POST action for the endpoint.
}

func myHandler(s state.State, r *http.Request) response.Response {
    msg := fmt.Sprintf("This is a response from %q at %q", s.Name(), s.Address())

    return response.SyncResponse(true, msg)

> }

Reviewer: Worth showing where to add the endpoint? I know the struct is shown above but maybe a comment can be provided that the endpoint is set as part of the `dargs`.

```

### Create your own schema extensions:
```go
var schemaUpdate1 schema.Update = func(ctx context.Context, tx *sql.Tx) error {
    _, err := tx.ExecContext(ctx, "my SQL statement")

    return err

> }

Reviewer: Same here let's have a comment telling that the actual schema update can be set in the `dargs`.

# Microcluster
## Introduction

> Microcluster is a Go library that provides an easy to use framework for creating and managing highly available [dqlite](https://github.com/canonical/dqlite) clusters. It provides an extensible API and database which can be used to directly connect to your underlying services.

Reviewer 2: ```suggestion
Microcluster is a Go library that provides an easy to use framework for creating and managing highly available [dqlite](https://github.com/canonical/dqlite) clusters (using [go-dqlite](https://github.com/canonical/go-dqlite)). It provides an extensible API and database which can be used to directly connect to your underlying services.
```

> Extend Microcluster with your own API endpoints, schema and API extensions, lifecycle actions, and additional network listeners.

Reviewer 2: "Extend Microcluster" sounds too gory; Microcluster gives me all of these tools for building my app. Take or leave:
```suggestion
Build your app with MicroCluster-managed HTTP endpoints, API extensions, database schema updates, lifecycle actions, and additional HTTP listeners.
```
Could also link to the structures/API for each of these things in the go doc.

## Import Microcluster

Get the latest LTS release of Microcluster:

```
$ go get github.com/canonical/microcluster/v2@latest
```

> ## Usage

Reviewer 2: How about "Building an app" or similar

Author: I went with `Configure and Start the Microcluster Service` because the example is just about starting the daemon and passing in external configuration.

For a step-by-step practical example of creating a Microcluster project, check out the [example package](example). Examine the code or build and test it out with the included Makefile.

All of Microcluster's state is stored in the `StateDir` directory. Define a new Microcluster app:

```go
m, err := microcluster.App(microcluster.Args{StateDir: "/path/to/state"})
if err != nil {
    // ...
}
```

> Extensions to Microcluster must be defined in `DaemonArgs`:

Reviewer 2: "Extensions" feels really overloaded to me; `APIExtensions` and `ExtensionsSchema` are _really_ different things. Could we say something like "Define application services with `DaemonArgs`"?

```go

dargs := microcluster.DaemonArgs{
    Version: "0.0.1",

    ExtensionsSchema: []schema.Update{
        // Sequential list of modifications to the database.
    },

    APIExtensions: []string{
        // Sequential list of API capabilities.
    },

    Servers: map[string]rest.Server{
        "my-server": {
            CoreAPI: true,
            ServeUnix: true,
            Resources: []rest.Resources{
                {
                    PathPrefix: "1.0",
                    Endpoints: []rest.Endpoint{
                        // Additional API endpoints to offer on "/1.0".
                    },
                },
            }
        }
    }
}
```

Start the daemon with your configuration:
```go
err := m.Start(ctx, dargs)
if err != nil {
    // ...
}
```

> ### Define lifecycle actions (Hooks):

Reviewer 2: "Where can I go to see all the hooks Microcluster provides?"

Maybe link [here](https://pkg.go.dev/github.com/canonical/microcluster@v0.0.0-20240731145923-49573627997a/internal/state#Hooks)?

Author: Linked to the code in the repo since it links to the `v3` branch head instead of a specific commit.

```go
dargs.Hooks = &state.Hooks{
    OnStart: func(ctx context.Context, s state.State) error {
        // Execute code on each startup.
    }
}
```

### Query the dqlite database directly:
```go

_, batch, err := m.Sql(ctx, "SELECT name, address FROM core_cluster_members WHERE role='voter'")
if err != nil {
    // ...
}

for i, result := range batch.Results {
    fmt.Printf("Query %d:\n", i)

    fmt.Println("Type:", result.Type)
    fmt.Println("Columns:", result.Columns)
    fmt.Println("RowsAffected:", result.RowsAffected)
    fmt.Println("Rows:", result.Rows)
}

```

### Create your own API endpoints:
```go
endpoint := rest.Endpoint{
    Path: "mypath" // API is served over /mypath

    Post: rest.EndpointAction{Handler: myHandler} // POST action for the endpoint.

func myHandler(s state.State, r *http.Request) response.Response {
    msg := fmt.Sprintf("This is a response from %q at %q", s.Name(), s.Address())

    return response.SyncResponse(true, msg)
}

> // Include your endpoints in DaemonArgs like so to serve over /1.0 over the default network.

Reviewer 2: ```suggestion
// Include your endpoints in DaemonArgs to serve over /1.0 over the default network.
```
I'm not a huge fan of "default network"; is there a name we can use to tie `CoreAPI` and `StartAPI`'s `newConfig.Location.Address` parameter? It took me a while to figure out what CoreAPI meant; this is a spot that could use some additional clarity IMO

Author: In MicroCloud we refer to this as the "cluster management network"

Reviewer 2: Yes, it's the default listener for the `/core/*` endpoints. I think I'd prefer at least "listener" rather than "network", since it's not really a "network" per se, just a handle to an addr+port listening for HTTP(S) traffic on some network interface.

dargs := microcluster.DaemonArgs{
    Servers: map[string]rest.Server{
        "my-server": {
            CoreAPI: true,
            ServeUnix: true,
            Resources: []rest.Resources{
                {
                    PathPrefix: "1.0",
                    Endpoints: []rest.Endpoint{endpoint},
                },
            }
        }
    }
}

```

### Create your own schema extensions:
```go
var schemaUpdate1 schema.Update = func(ctx context.Context, tx *sql.Tx) error {

>     _, err := tx.ExecContext(ctx, "my SQL statement")

Reviewer 2: One-liner `CREATE TABLE`?

# Microcluster
## Introduction

> Microcluster is a Go library that provides an easy to use framework for creating and managing highly available [dqlite](https://github.com/canonical/dqlite) clusters (using [go-dqlite](https://github.com/canonical/go-dqlite)). It provides an extensible API and database which can be used to directly connect to your underlying services.

Reviewer 3: ```suggestion
Microcluster is a Go library that provides an easy-to-use framework for creating and managing highly available [Dqlite](https://github.com/canonical/dqlite) clusters (using [go-dqlite](https://github.com/canonical/go-dqlite)). It offers an extensible API and database, which can be used to directly connect to your underlying services.
```

Build your service with Microcluster-managed HTTPS endpoints, API extensions, schema updates, lifecycle actions, and additional HTTPS listeners.

## Import Microcluster

Get the latest LTS release of Microcluster:

```
$ go get github.com/canonical/microcluster/v2@latest
```

> ## Configure and Start the Microcluster Service

Reviewer 3: ```suggestion
## Configure and start the Microcluster service
```

> For a step-by-step practical example of creating a Microcluster project, check out the [example package](example). Examine the code or build and test it out with the included Makefile.

Reviewer 2: ```suggestion
For a step-by-step practical example of creating a Microcluster project, check out the [example package](example). Examine the code or build and test it with the included Makefile.
```

All of Microcluster's state is stored in the `StateDir` directory. Define a new Microcluster app:

```go
m, err := microcluster.App(microcluster.Args{StateDir: "/path/to/state"})
if err != nil {
    // ...
}
```

Define application services for Microcluster in `DaemonArgs`:
```go

dargs := microcluster.DaemonArgs{
    Version: "0.0.1",

    ExtensionsSchema: []schema.Update{
        // Sequential list of modifications to the database.
    },

    APIExtensions: []string{
        // Sequential list of API capabilities.
    },

    Servers: map[string]rest.Server{
        "my-server": {
            CoreAPI: true,
            ServeUnix: true,
            Resources: []rest.Resources{
                {
                    PathPrefix: "1.0",
                    Endpoints: []rest.Endpoint{
                        // Additional API endpoints to offer on "/1.0".
                    },
                },
            }
        }
    }
}
```

Start the daemon with your configuration:
```go
err := m.Start(ctx, dargs)
if err != nil {
    // ...
}
```

> ### Define lifecycle actions (Hooks):

Reviewer 3: ```suggestion
### Define lifecycle actions (hooks)
```

> The complete set of hooks and their behaviours are defined [here](https://github.com/canonical/microcluster/blob/v3/internal/state/hooks.go)

Reviewer 3: ```suggestion
The complete set of Microcluster hooks and their behaviors are defined [in this Go file](https://github.com/canonical/microcluster/blob/v3/internal/state/hooks.go).
```

```go
dargs.Hooks = &state.Hooks{
    OnStart: func(ctx context.Context, s state.State) error {
        // Execute code on each startup.
    }
}
```

> ### Query the dqlite database directly:

Reviewer 3: ```suggestion
### Query the Dqlite database directly
```

```go

_, batch, err := m.Sql(ctx, "SELECT name, address FROM core_cluster_members WHERE role='voter'")
if err != nil {
    // ...
}

for i, result := range batch.Results {
    fmt.Printf("Query %d:\n", i)

    fmt.Println("Type:", result.Type)
    fmt.Println("Columns:", result.Columns)
    fmt.Println("RowsAffected:", result.RowsAffected)
    fmt.Println("Rows:", result.Rows)
}

```

> ### Create your own API endpoints:

Reviewer 3: ```suggestion
### Create your own API endpoints
```

```go
endpoint := rest.Endpoint{
    Path: "mypath" // API is served over /mypath

    Post: rest.EndpointAction{Handler: myHandler} // POST action for the endpoint.

func myHandler(s state.State, r *http.Request) response.Response {
    msg := fmt.Sprintf("This is a response from %q at %q", s.Name(), s.Address())

    return response.SyncResponse(true, msg)
}

// Include your endpoints in DaemonArgs like so to serve over /1.0 over the default listener.
dargs := microcluster.DaemonArgs{
    Servers: map[string]rest.Server{
        "my-server": {
            CoreAPI: true,
            ServeUnix: true,
            Resources: []rest.Resources{
                {
                    PathPrefix: "1.0",
                    Endpoints: []rest.Endpoint{endpoint},
                },
            }
        }
    }
}

```

> ### Create your own schema extensions:

Reviewer 3: ```suggestion
### Create your own schema extensions
```
